### PR TITLE
Build: Migrate unit tests from Jest to Vitest

### PR DESCRIPTION
**Pull request by yannbf**
## What I did

This migrates the monorepo unit tests from Jest to Vitest.  This will make it easier to convert our codebase to ESM, and vitest has some good DX which we can take advantage of.

I wrote a codemod to do most of the manual labor and you can find it here: https://go.intuita.io/IX3EFT

<details>
<summary>The codemod takes care of:</summary>

1. JSDOM environment comments
```ts
// from
/**
 * @jest-environment jsdom
 */

// to
/**
 * @vitest-environment jsdom
 */
```

2. Type reference comments
```ts
// from
/// <reference types="@types/jest" />;

// to
/// <reference types="vitest" />;
```

3. All sorts of spy/mock/fn migrations
```ts
// from
jest.mock('@storybook/preview-api');
const value = jest.requireActual('@storybook/preview-api');
const value2 = jest.createMockFromModule('@storybook/preview-api');
const myMock = jest.fn();
const spy = jest.spyOn(EventsPackageExport, 'bla');

// to
import { vi } from "vitest";

vi.mock('@storybook/preview-api');
const value = await vi.importActual('@storybook/preview-api');
const value2 = vi.mock('@storybook/preview-api')
const myMock = vi.fn();
const spy = vi.spyOn(EventsPackageExport, 'bla');
```

4. All sorts of TypeScript type migrations
```ts
// from
const mockedApi = api as unknown as jest.Mocked<api.API>;
const toIdMock = toId as jest.Mock<ReturnType<typeof toId>>;
let mockChannel: { on: jest.Mock; emit?: jest.Mock };
(helpers.isNxProject as jest.SpyInstance).mockResolvedValue(true);

// to
import { Mock, Mocked, SpyInstance } from "vitest";

const mockedApi = api as unknown as Mocked<api.API>;
const toIdMock = toId as Mock<ReturnType<typeof toId>>;
let mockChannel: { on: Mock; emit?: Mock };
(helpers.isNxProject as SpyInstance).mockResolvedValue(true);
```

Apart from adding missing imports for `describe, it, test, beforeEach, beforeAll, afterEach, afterAll`, etc.

</details>

To sum the changes up:
- Codemod ran across all files in the codebase
- storyshots and storyshots-puppeteer packages are deleted. (they are deprecated, and in Storybook 8 they would be removed. Else we'd need to do a rework on them for proper ESM support)
- Mocking mechanism changed e.g.
- A lot of snapshots are updated
- Some package.json exports were changed
- jest config files were turned into vitest config files
- `jest-os-detection` (for windowsOnly tests) usage was replaced with a small local utility
- `jest-specific-snapshot` was replaced with Vitest's own snapshot utility

## Solved problems

We encountered and overcame some challenges in this conversion, see collapsed details below if you're interested.

<details>
<summary>details</summary>

### Out of memory issue - Solved!

~When running the tests, the process hangs for about 3min and then presents the following error message:
image~

~This has to be investigated and fixed.~

Fixed by https://github.com/storybookjs/storybook/pull/24415/commits/6d35099f00af18056ec2dc01537b237f4acff3da

### Emotion loading twice - Solved!

It seems that @emotion/react is being loaded in two different contexts, which causes the theme to be lost, and some tests to fail. It's unclear what is causing this.  This is breaking tests in `addons/interactions`, `addons/a11y`, and `ui/blocks`, and `ui/manager`.

### Call counts in StoryIndexGenerator tests are off by one - Solved!

`AssertionError: expected "N" to be called 6 times, but got 5 times`.  It's unclear what caused this, or whether it is truly broken.  This breaks tests in `lib/core-server`.

</details>

### Skipped tests

We couldn't figure out an emotion theming issue for a couple of tests, so chose to temporarily skip them to unblock the rest of the conversion.

- `addons/a11y/src/components/A11YPanel.test.tsx`
- `addons/a11y/src/components/VisionSimulator.test.tsx`

#### The changes in this PR are covered in the following automated tests:
- [x] stories
- [x] unit tests
- [x] integration tests
- [ ] end-to-end tests

#### Manual testing

Run `yarn test` or `yarn test:watch` in the `code` directory.

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

**Comment by IanVS** [comment]
Status update: I've gotten the `renderers` passing, and most of `addons`, except for a problem whereby emotion is somehow loaded twice, so the test loses the `theme` context.

File: code/renderers/vue/package.json [hunk 1]
```diff
@@ -77,6 +79,9 @@
   "engines": {
     "node": ">=16.0.0"
   },
+  "installConfig": {
+    "hoistingLimits": "workspaces"
```
Comment (IanVS) [inline code/renderers/vue/package.json:83]: I needed to add this to keep `vue-template-compiler` from being hoisted up to the monorepo root, where it would find vue3 (also hoisted) and crash.

File: code/lib/core-common/src/presets.ts [hunk 2]
```diff
@@ -210,8 +209,7 @@ async function getContent(input: any) {
     return rest;
   }
   const name = input.name ? input.name : input;
-
-  return interopRequireDefault(name);
+  return import(name);
```
Comment (IanVS) [inline code/lib/core-common/src/presets.ts:212]: @ndelangen this is a bit of a risky change, and needs your 👀.

I do know that without this, it's going to be very difficult to use vitest, since it only mocks `import` and not `require`

File: code/addons/interactions/package.json [hunk 3]
```diff
@@ -43,6 +43,11 @@
       "require": "./dist/manager.js",
       "import": "./dist/manager.mjs"
     },
+    "./preset": {
+      "types": "./dist/preset.d.ts",
+      "require": "./dist/preset.js",
+      "import": "./dist/preset.js"
```
Comment (IanVS) [inline code/addons/interactions/package.json:49]: This `import` condition and the one above it are not correct, they're pointing to cjs files.  But we don't generate an ESM version, so I did this just to get tests passing.  It should probably be fixed the right way by generating esm.

File: code/lib/core-common/src/presets.test.ts [hunk 4]
```diff
@@ -26,7 +41,7 @@ vi.mock('./utils/safeResolve', () => {
   const KNOWN_FILES = [
     '@storybook/react',
     '@storybook/addon-actions/manager',
-    '@storybook/addon-actions/register',
+    '@storybook/addon-actions/register.js',
```
Comment (IanVS) [inline code/lib/core-common/src/presets.test.ts:44]: We don't have a condition for `/register`, only `./register.js`

File: code/lib/theming/src/utils.ts [hunk 5]
```diff
@@ -63,9 +60,9 @@ export const darkenColor = colorFactory('darken');
 // The default color scheme is light so unless the preferred color
 // scheme is set to dark we always want to use the light theme
 export const getPreferredColorScheme = () => {
-  if (!globalWindow || !globalWindow.matchMedia) return 'light';
+  if (!globalThis || !globalThis.matchMedia) return 'light';
```
Comment (IanVS) [inline code/lib/theming/src/utils.ts:63]: I believe `globalThis` is supported in all node and browser versions that we are currently targeting, and it makes mocking in the tests much easier.

File: code/lib/theming/src/tests/util.test.js [hunk 6]
```diff
@@ -100,9 +100,12 @@ describe('utils', () => {
...
       // is a simple way to test it
-      globalWindow.matchMedia = vi.fn().mockImplementation(() => ({
-        matches: true,
-      }));
+      vi.stubGlobal(
+        'matchMedia',
+        vi.fn().mockReturnValue({
+          matches: true,
+        })
+      );
```
Comment (IanVS) [inline code/lib/theming/src/tests/util.test.js:108]: We can also just use `globalThis.matchMedia = ` directly, but then we can't use `vi.unstubAllGlobals` if we need it.  (ref https://vitest.dev/api/vi.html#vi-stubglobal)

**Comment by IanVS** [comment]
I've done what I can for now, I need to step back and work on other things, but hopefully someone else can pick up the mantle and get this across the finish line.

**Comment by JReinhold** [comment]
I've investigated and fixed the tests related to `StoryIndexGenerator`.
The tests mocked a function in a separate package, and counted how many times that mock was called.

In Jest the mock was applied both in the module under test, but also any sub-packages that were being invoked had the mock as well. In Vitest however, sub-packages that use the same mocked package for some reason don't get the mock function, they still call the real function.

This caused the mock call count to be off by one between Jest and Vitest. I changed the expected count, because the count is not actually what's important in the tests, so the tests still asserts what's needed.

**Comment by ndelangen** [comment]
@SocketSecurity ignore @vitest/snapshot@0.34.6
@SocketSecurity ignore @vitest/snapshot@1.0.0-beta.3
@SocketSecurity ignore vue-template-compiler@2.6.14
@SocketSecurity ignore tinypool@0.8.1
@SocketSecurity ignore vitest@0.34.6
@SocketSecurity ignore vitest@1.0.0-beta.3

**Comment by ndelangen** [comment]
I tried... 😢

**Comment by IanVS** [comment]
@ndelangen what is left to fix here?

**Comment by JReinhold** [comment]
@IanVS I've updated the list of failing tests. It looks like basing this on `release-8-0` instead of `next` has caused more tests to fail with the double-emotion loading error, most likely because of the new layout architecture.

**Comment by JReinhold** [comment]
I've fixed the emotion double loading error. This hint was here:

https://github.com/emotion-js/emotion/pull/2905

This is a known issue in emotion, and they solve it by checking for the `vi` global before showing the warning. However [Vitest doesn't include that global by default](https://vitest.dev/config/#globals), but including it fixed the issue. (I don't think we should use the global, so I haven't configured TS to know about it)

**Comment by IanVS** [comment]
I believe this is now ready for review.  

image

I left a few comments above for some of the trickier or potentially problematic changes I made, which may now be collapsed.  So whoever reviews this, please give those a read.

**Comment by IanVS** [comment]
Ergh, looks like there are still some CI failures.  But we're 🤏 close.

File: code/renderers/vue/package.json (hunk 1 above)
Comment (JReinhold) [reply code/renderers/vue/package.json:81]: We're removing Vue2 in SB 8 anyway, so this doesn't matter.

File: code/lib/core-common/src/presets.ts (hunk 2 above)
Comment (JReinhold) [reply code/lib/core-common/src/presets.ts]: I've managed to revert this in https://github.com/storybookjs/storybook/pull/24415/commits/6dfcad5214340af3080d958777599aaa5efed612 and use `mock-require` to mock the presets instead. Seems to work

File: code/lib/theming/src/utils.ts (hunk 5 above)
Comment (JReinhold) [reply code/lib/theming/src/utils.ts:63]: It would be interesting to try and replace `@storybook/global` with `globalThis` everywhere someday.

**Review by JReinhold — CHANGES_REQUESTED** [review]
I went through the whole thing (except for the gazillion jest -> vitest config changes TBH) and it looks great.

I only have a blocker on the snapshots that don't look right, I'll investigate it.

File: code/renderers/svelte/src/__test__/Button.svelte [hunk 7]
```diff
@@ -1,8 +1,13 @@
 <script lang="ts">
-  import { createEventDispatcher } from 'svelte';
+  const { createEventDispatcher } = require('svelte');
```
Comment (IanVS) [inline code/renderers/svelte/src/__test__/Button.svelte:2]: This seems strange, I thought svelte was all-in on ESM.  Why was this change needed?

Comment (IanVS) [reply code/renderers/svelte/src/__test__/Button.svelte:2]: This is what I did previously to get those tests passing: https://github.com/storybookjs/storybook/pull/24415/commits/ca0394d77c0d41009181e73848634cdd6a3bd3c8

I see now there's an `// eslint-disable-next-line import/no-unresolved` comment above the import, so maybe that has something to do with it?  Why wouldn't that import be resolved?  And if it's not, it would explain why the files can't be parsed.

File: code/lib/core-common/src/utils/__tests__/normalize-stories.test.ts [hunk 8]
```diff
@@ -17,7 +18,7 @@ expect.addSnapshotSerializer({
 });
 
 expect.extend({
-  toMatchPaths(regex: RegExp, paths: string[]) {
+  toEqual(regex: RegExp, paths: string[]) {
```
[the remark this answers is not in the document]
Comment (yannbf) [reply code/lib/core-common/src/utils/__tests__/normalize-stories.test.ts:21]: fixed!

File: code/renderers/svelte/src/__test__/Button.svelte (hunk 7 above)
Comment (yannbf) [reply code/renderers/svelte/src/__test__/Button.svelte]: Seems like this was not needed after all, it was causing issues which got solved after undoing these changes

File: code/lib/core-common/vitest.d.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable @typescript-eslint/no-empty-interface */
+import 'vitest';
+
+interface CustomMatchers<R = unknown> {
+  toMatchPaths(paths: string[]): R;
```
Comment (IanVS) [inline code/lib/core-common/vitest.d.ts:5]: The `expect.extend` should probably move to a test-setup file so it's available for every test, if we're adding the types like this.  It would be nice if there was a way to handle the types for an extends in just one test, though.

File: code/renderers/react/src/docs/jsxDecorator.test.tsx [hunk 10]
```diff
@@ -120,10 +122,10 @@ describe('renderJsx', () => {
 
     expect(renderJsx(createElement(MyExoticComponent, {}, 'I am forwardRef!'), {}))
       .toMatchInlineSnapshot(`
-      <MyExoticComponent>
-        I am forwardRef!
-      </MyExoticComponent>
-    `);
+        <MyExoticComponent2>
+          I am forwardRef!
+        </MyExoticComponent2>
```
[the remark this answers is not in the document]
Comment (yannbf) [reply code/renderers/react/src/docs/jsxDecorator.test.tsx:127]: it's because the component is called MyExoticComponent and the function which is forwarded is also called MyExoticComponent

File: code/lib/core-common/vitest.d.ts (hunk 9 above)
Comment (yannbf) [reply code/lib/core-common/vitest.d.ts:5]: sounds good to me!

**Review by ndelangen — APPROVED** [review]
